MQClient/backends/pulsar.py

```diff
@@ -0,0 +1,189 @@
+    return q
+
+
+def send_message(queue: PulsarPub, msg: bytes) -> None:
+    """Send a message on a queue."""
+    if not queue.producer:
+        raise RuntimeError("queue is not connected")
+
+    queue.producer.send(msg)
+
+
+def get_message(queue: PulsarSub, timeout_millis: int = 100) -> typing.Optional[Message]:
+    """Get a single message from a queue.
+
+    To endlessly block until a message is available, set `timeout_millis=None`.
+    """
+    if not queue.consumer:
+        raise RuntimeError("queue is not connected")
+
+    while True:
+        try:
+            msg = queue.consumer.receive(timeout_millis=timeout_millis)
+            if msg:
+                message_id, data = msg.message_id(), msg.data()
+                if message_id and data:
+                    return Message(message_id, data)
+            return None
+
+        except Exception as e:
+            if str(e) == "Pulsar error: TimeOut":  # pulsar isn't a fan of derived Exceptions
+                return None
+            if str(e) == "Pulsar error: AlreadyClosed":
+                queue.connect()
+                continue
+            raise
+
+        break  # poor man's do-while loop
+
+    raise Exception('Pulsar connection error')
+
+
+def ack_message(queue: PulsarSub, msg_id: MessageID) -> None:
+    """Ack a message from the queue."""
+    if not queue.consumer:
+        raise RuntimeError("queue is not connected")
+
+    queue.consumer.acknowledge(msg_id)
+
+
+def reject_message(queue: PulsarSub, msg_id: MessageID) -> None:
+    """Reject (nack) a message from the queue."""
+    if not queue.consumer:
+        raise RuntimeError("queue is not connected")
+
+    queue.consumer.negative_acknowledge(msg_id)
+
+
+def message_generator(queue: PulsarSub, timeout: int = 60, auto_ack: bool = True,
+                      propagate_error: bool = True) -> typing.Generator[Message, None, None]:
+    """Yield Messages.
+
+    Arguments:
+        queue {PulsarSub} -- queue object
+
+    Keyword Arguments:
+        timeout {int} -- timeout in seconds for inactivity (default: {60})
+        auto_ack {bool} -- Ack each message after successful processing (default: {True})
+        propagate_error {bool} -- should errors from downstream code kill the generator? (default: {True})
+    """
+    if not queue.consumer:
+        raise RuntimeError("queue is not connected")
+
+    try:
+        while True:
+            msg = get_message(queue, timeout_millis=timeout * 1000)
+            if not msg:
+                logging.info("no messages in idle timeout window")
+                break
+            try:
+                yield msg
+            except Exception as e:  # pylint: disable=W0703
+                reject_message(queue, msg.msg_id)
+                if propagate_error:
+                    raise
+                logging.warning('error downstream: %r', e, exc_info=True)
+            else:
+                if auto_ack:
+                    ack_message(queue, msg.msg_id)
```

Comment:
This seems a little odd; can `yield msg` actually raise an Exception?

Reply:
This is a generator, so the `yield msg` is passing control to somewhere else, which may raise an Exception.

Reply:
I think we're missing a unit test that triggers this though.

Reply:
Trying to make a code snippet that does this, I may have gotten generators and context managers mixed together.  It doesn't seen like generators can catch exceptions.

So, we probably need to change how this function works.  But I recommend another pull request for that.

Reply:
This function was essentially lifted straight from RabbitMQ. So, it makes sense to punt this